implementations/rust/ockam/ockam_app_lib/src/incoming_services/state.rs

```diff
@@ -0,0 +1,365 @@
+    }
+
+    pub fn enabled(&self) -> bool {
+        self.enabled
+    }
+
+    pub fn set_port(&mut self, port: Option<u16>) {
+        self.port = port;
+    }
+
+    pub fn enable(&mut self) {
+        self.enabled = true;
+    }
+    pub fn disable(&mut self) {
+        self.enabled = false;
+        self.port = None;
+    }
+
+    /// Returns the enrollment ticket, to avoid conflicts with 'default' name
+    /// the project name is re-set to the project id
+    pub fn enrollment_ticket(&self) -> ockam::Result<EnrollmentTicket> {
+        let mut enrollment_ticket = self.access_details.enrollment_ticket()?;
+        if let Some(project) = enrollment_ticket.project.as_mut() {
+            project.name = project.id.clone();
+        }
+        Ok(enrollment_ticket)
+    }
+
+    pub fn relay_name(&self) -> String {
+        let bare_relay_name = self.access_details.shared_node_identity.to_string();
+        format!("forward_to_{bare_relay_name}")
+    }
+
+    /// Returns the name of the inlet within the node, for now it's a constant
+    pub fn inlet_name(&self) -> &str {
+        "app-inlet"
+    }
+
+    pub fn service_route(&self) -> ockam::Result<String> {
+        if let Some(project) = self.enrollment_ticket()?.project.as_ref() {
+            let project_id = &project.id;
+            let relay_name = self.relay_name();
+            let service_name = match self.original_name() {
+                Ok(name) => name,
+                Err(_) => {
+                    warn!("Failed to get service name from access details");
+                    return Err(ApiError::core(
+                        "Failed to get service name from access details",
+                    ));
+                }
+            };
+            Ok(format!(
+                "/project/{project_id}/service/{relay_name}/secure/api/service/{service_name}"
+            ))
+        } else {
+            Err(ApiError::core("No project id found in enrollment ticket"))
+        }
+    }
+
+    pub fn local_node_name(&self) -> ockam::Result<String> {
+        if let Some(project) = self.enrollment_ticket()?.project.as_ref() {
+            let project_id = &project.id;
+            let id = &self.id;
+            Ok(format!("ockam_app_{project_id}_{id}"))
+        } else {
+            Err(ApiError::core("No project id found in enrollment ticket"))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::incoming_services::PersistentIncomingServiceState;
+    use crate::state::AppState;
+    use ockam::identity::{Identifier, OneTimeCode};
+    use ockam_api::cli_state::CliState;
+    use ockam_api::cloud::share::{
+        InvitationWithAccess, ReceivedInvitation, RoleInShare, ServiceAccessDetails, ShareScope,
+    };
+    use ockam_api::config::lookup::ProjectLookup;
+    use ockam_api::identity::EnrollmentTicket;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_inlet_data_from_invitation() {
+        let mut invitation = InvitationWithAccess {
```

Comment:
This could be moved closer to L.265. I would also advise to factor out the code creating data outside of the test as utility functions so that:

 - the structure of the test become clearer
 - we can reuse those functions for other tests